File: QueueUsingStacks.cs
// Time Complexity : O(1)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No

using System;
using System.Collections.Generic;

public class MyQueue
{
    private Stack<int> inSt;
    private Stack<int> outSt;

    public MyQueue()
    {
        inSt = new Stack<int>();
        outSt = new Stack<int>();
    }

    public void Push(int x)
    {
        inSt.Push(x);
    }

    public int Pop()
    {
        if (outSt.Count == 0)
        {
            while (inSt.Count > 0)
            {
                outSt.Push(inSt.Pop());
            }
        }
        return outSt.Pop();
    }

    public int Peek()
    {
        if (outSt.Count == 0)
        {
            while (inSt.Count > 0)
            {
                outSt.Push(inSt.Pop());
            }
        }
        return outSt.Peek();
    }

    public bool Empty()
    {
        return inSt.Count == 0 && outSt.Count == 0;
    }
}

// Main program to run the test case
public class Program
{
    public static void Main(string[] args)
    {
        // Example input as per the problem:
        // ["MyQueue", "push", "push", "peek", "pop", "empty"]
        // [[], [1], [2], [], [], []]

        MyQueue obj = new MyQueue();
        Console.WriteLine("null");  // Corresponds to the instantiation of the MyQueue object

        obj.Push(1);
        Console.WriteLine("null");  // Corresponds to the first push operation

        obj.Push(2);
        Console.WriteLine("null");  // Corresponds to the second push operation

        int param_3 = obj.Peek();
        Console.WriteLine(param_3);  // Expected output: 1 (peek operation)

        int param_2 = obj.Pop();
        Console.WriteLine(param_2);  // Expected output: 1 (pop operation)

        bool param_4 = obj.Empty();
        Console.WriteLine(param_4);  // Expected output: False (empty operation)
    }
}

File: queueUsingStacks.cpp
// Time Complexity : 
// 1. push(): O(1)
// 2. pop(): O(n) - n is the number of elements present in the stack
// 3. peek(): O(1)
// 4. empty(): O(1)
// Space Complexity : O(n) - n is the number of elements present in the stack

// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach

class MyQueue {
    stack<int> s1;
    stack<int> s2;
    int front;
public:
    MyQueue() {
        
    }
    
    // if s1 is empty we assign the first element as top
    // we push it to stack 1
    void push(int x) {
        if (s1.empty()) {
            front = x;
        }

        s1.push(x);
    }
    
    // transfer elements from s1 to s2 and return the top element of s2
    int pop() {
        if (s2.empty()) {
            while(!s1.empty()) {
                int top = s1.top();
                s1.pop();
                s2.push(top);
            }
        }
        int top = s2.top();
        s2.pop();

        return top;
    }
    
    // if s2 is empty return front else return front.
    int peek() {
        if(!s2.empty()) {
            return s2.top();
        }

        return front;
    }
    
    // check both the stacks, if both of them are empy return true
    bool empty() {
        if (s1.empty() && s2.empty()) {
            return true;
        }

        return false;
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */